### PR TITLE
Matrix3 implementation

### PR DESCRIPTION
**Pull request by GonzaCerv**

### CODE REVIEW

File: course/include/Isometry.hpp [hunk 1]
```diff
@@ -91,5 +92,44 @@ inline std::ostream& operator<<(std::ostream& os, const Vector3& vect) {
...
+                values_.push_back(m10);
+                values_.push_back(m11);
+                values_.push_back(m13);
+                values_.push_back(m21);
+                values_.push_back(m22);
+                values_.push_back(m23);
+            }
+    Matrix3(const std::initializer_list<double>& row_1, const std::initializer_list<double>& row_2,
+            const std::initializer_list<double>& row_3){
+                for(auto i : row_1){
```
Comment (glpuga) [inline course/include/Isometry.hpp]: nit: reserve the `i` name for counters. It's so conventional that people expect it to be like that. Call it `item`, or `cell`, or `number` .

Comment (glpuga) [inline course/include/Isometry.hpp]: Maybe you should check the lengths of the initializer lists before.

File: course/src/Isometry.cc [hunk 2]
```diff
@@ -92,5 +94,23 @@ const Vector3 Vector3::kUnitY(0.0, 1.0, 0.0);
...
 
+// Matrix3
+
+const Matrix3 Matrix3::kIdentity({1.0, 0.0, 0.0}, {0.0, 1.0, 0.0}, {0.0, 0.0, 1.0});
+const Matrix3 Matrix3::kZero({0.0, 0.0, 0.0}, {0.0, 0.0, 0.0}, {0.0, 0.0, 0.0});
+
+bool Matrix3::operator==(const Matrix3& a) const {
+    auto it = values_.begin();
+    auto a_it = a.values_.begin();
+    for (uint8_t count = 0; count < values_.size(); ++count) {
```
Comment (glpuga) [inline course/src/Isometry.cc]: What happens if both matrices don't have the same number of elements? Code defensively.

Comment (GonzaCerv) [reply course/src/Isometry.cc]: If I pass a Matrix3 as a parameter, shouldn't be always initialized?

File: course/include/Isometry.hpp (hunk 1 above)
Comment (GonzaCerv) [reply course/include/Isometry.hpp]: Done!

Comment (GonzaCerv) [reply course/include/Isometry.hpp]: Done!

**Comment by GonzaCerv** [comment]
@glpuga this PR is ready for review

File: course/src/Isometry.cc (hunk 2 above)
Comment (glpuga) [reply course/src/Isometry.cc]: Now they are, but before this you were not checking the length of the initliazer list constructor.

File: course/include/Isometry.hpp [hunk 3]
```diff
@@ -93,43 +93,64 @@ inline std::ostream& operator<<(std::ostream& os, const Vector3& vect) {
 }
 
 class Matrix3 {
+    using mat3 = std::vector<std::vector<double>>;
```
Comment (glpuga) [inline course/include/Isometry.hpp]: Better use three `Vector3` as rows (or columns). That way you can levarage the operations  you already implemented for Vector3 for Matrix3.

File: course/include/Isometry.hpp [hunk 4]
```diff
@@ -93,43 +93,64 @@ inline std::ostream& operator<<(std::ostream& os, const Vector3& vect) {
...
-                }
-                for(auto i : row_2){
-                    values_.push_back(i);
-                }
-                for(auto i : row_3){
-                    values_.push_back(i);
-                }
-            }
+            const std::initializer_list<double>& row_3);
+    Matrix3();
```
Comment (glpuga) [inline course/include/Isometry.hpp:102]: For the sake of practice also create a copy constructor and a move constructor.

File: course/include/Isometry.hpp [hunk 5]
```diff
@@ -93,43 +93,64 @@ inline std::ostream& operator<<(std::ostream& os, const Vector3& vect) {
...
 
     // Operators
     bool operator==(const Matrix3& a) const;
-    
+    bool operator==(const std::initializer_list<double>& rhs) const;
+    Matrix3 operator+(const Matrix3& a) const;
+    Matrix3 operator-(const Matrix3& a) const;
```
Comment (glpuga) [inline course/include/Isometry.hpp:111]: For the sake of practice also create a move assignment operator and a copy assignment operator.

File: course/include/Isometry.hpp [hunk 6]
```diff
@@ -93,43 +93,64 @@ inline std::ostream& operator<<(std::ostream& os, const Vector3& vect) {
...
+    Matrix3 operator*(const double& a) const;
+    Matrix3 operator*(const Matrix3& a) const;
+    Matrix3 operator/(const Matrix3& a) const;
+    const std::vector<double>& operator[](int val) const;
```
Comment (glpuga) [inline course/include/Isometry.hpp]: This should return a Vector3.

File: course/include/Isometry.hpp [hunk 7]
```diff
@@ -93,43 +93,64 @@ inline std::ostream& operator<<(std::ostream& os, const Vector3& vect) {
...
+    std::vector<double>& operator[](int val);
```
Comment (glpuga) [inline course/include/Isometry.hpp]: Same.

File: course/include/Isometry.hpp [hunk 8]
```diff
@@ -93,43 +93,64 @@ inline std::ostream& operator<<(std::ostream& os, const Vector3& vect) {
...
+
+    // Methods
+    double det();
```
Comment (glpuga) [inline course/include/Isometry.hpp]: Will you be able to calculate the determinant of an rvalue or a cost variable like this?

File: course/include/Isometry.hpp [hunk 9]
```diff
@@ -93,43 +93,64 @@ inline std::ostream& operator<<(std::ostream& os, const Vector3& vect) {
...
-    std::vector<double> values_;
-
+    mat3 mat_;
 };
 
+// non-member binary operators
+
+inline Matrix3 operator*(const double scalar, const Matrix3& rhs) {
+    Matrix3 result;
+    for (uint32_t y = 0; y < 3; ++y) {
```
Comment (glpuga) [inline course/include/Isometry.hpp]: Store the matrix as vectors, and use the vector operators to implement this.

File: course/src/Isometry.cc [hunk 10]
```diff
@@ -96,21 +98,154 @@ const Vector3 Vector3::kZero(0.0, 0.0, 0.0);
...
+                 const std::initializer_list<double>& row_3) {
+    mat_.resize(3);
+    for (unsigned item = 0; item < 3; item++) {
+        mat_[item].resize(3, 0);
+    }
+    if ((row_1.size() < 3) || (row_2.size() < 3) || (row_3.size() < 3)) {
+        return;
+    }
+    auto it = row_1.begin();
+    mat_[0][0] = *it;
```
Comment (glpuga) [inline course/src/Isometry.cc]: Could have been `mat_[0][0] = *it++;`

File: course/src/Isometry.cc [hunk 11]
```diff
@@ -96,21 +98,154 @@ const Vector3 Vector3::kZero(0.0, 0.0, 0.0);
...
-const Matrix3 Matrix3::kIdentity({1.0, 0.0, 0.0}, {0.0, 1.0, 0.0}, {0.0, 0.0, 1.0});
-const Matrix3 Matrix3::kZero({0.0, 0.0, 0.0}, {0.0, 0.0, 0.0}, {0.0, 0.0, 0.0});
+Matrix3::Matrix3(const std::initializer_list<double>& row_1, const std::initializer_list<double>& row_2,
+                 const std::initializer_list<double>& row_3) {
+    mat_.resize(3);
+    for (unsigned item = 0; item < 3; item++) {
+        mat_[item].resize(3, 0);
+    }
+    if ((row_1.size() < 3) || (row_2.size() < 3) || (row_3.size() < 3)) {
+        return;
```
Comment (glpuga) [inline course/src/Isometry.cc]: Better throw an error.

File: course/src/Isometry.cc [hunk 12]
```diff
@@ -96,21 +98,154 @@ const Vector3 Vector3::kZero(0.0, 0.0, 0.0);
...
 
 bool Matrix3::operator==(const Matrix3& a) const {
-    auto it = values_.begin();
-    auto a_it = a.values_.begin();
-    for (uint8_t count = 0; count < values_.size(); ++count) {
-        if (!almost_equal(*a_it, *it, resolution)) {
-            return false;
+    for (auto x = 0; x < 3; ++x) {
+        for (auto y = 0; y < 3; ++y) {
+            if (!almost_equal(mat_[x][y], a[x][y], resolution)) {
```
Comment (glpuga) [inline course/src/Isometry.cc]: Use vector operators.

File: course/src/Isometry.cc [hunk 13]
```diff
@@ -96,21 +98,154 @@ const Vector3 Vector3::kZero(0.0, 0.0, 0.0);
...
 }
 
+bool Matrix3::operator==(const std::initializer_list<double>& rhs) const {
+    if (rhs.size() < 9) {
+        return false;
+    }
+    auto it = rhs.begin();
+    for (auto x = 0; x < 3; ++x) {
+        for (auto y = 0; y < 3; ++y) {
+            if (!almost_equal(mat_[x][y], *it, resolution)) {
```
Comment (glpuga) [inline course/src/Isometry.cc]: Same.

File: course/src/Isometry.cc [hunk 14]
```diff
@@ -96,21 +98,154 @@ const Vector3 Vector3::kZero(0.0, 0.0, 0.0);
...
+            ++it;
+        }
+    }
+    return true;
+}
+
+Matrix3 Matrix3::operator+(const Matrix3& a) const {
+    Matrix3 result;
+    for (uint32_t x = 0; x < 3; ++x) {
+        for (uint32_t y = 0; y < 3; ++y) {
```
Comment (glpuga) [inline course/src/Isometry.cc]: Same in all operators below.

File: course/src/Isometry.cc [hunk 15]
```diff
@@ -96,21 +98,154 @@ const Vector3 Vector3::kZero(0.0, 0.0, 0.0);
...
+}
+
+const std::vector<double>& Matrix3::operator[](int val) const {
+    if (val > 3) {
+        return mat_[3];
+    }
+    return mat_[val];
+}
+
+std::vector<double>& Matrix3::operator[](int val) { return mat_[val]; }
```
Comment (glpuga) [inline course/src/Isometry.cc]: Check the range of `val`.

File: course/src/Isometry.cc [hunk 16]
```diff
@@ -96,21 +98,154 @@ const Vector3 Vector3::kZero(0.0, 0.0, 0.0);
...
+    for (uint32_t x = 0; x < 3; ++x) {
+        for (uint32_t y = 0; y < 3; ++y) {
+            result[x][y] = mat_[x][y] / a[x][y];
+        }
+    }
+    return result;
+}
+
+const std::vector<double>& Matrix3::operator[](int val) const {
+    if (val > 3) {
```
Comment (glpuga) [inline course/src/Isometry.cc]: Check the range of `val`.

**Review by glpuga — CHANGES_REQUESTED** [review]
Good work, but we can reduce the work reusing `Vector3`.

**Comment by GonzaCerv** [comment]
Comments addressed!

File: course/src/Isometry.cc [hunk 17]
```diff
@@ -96,21 +98,152 @@ const Vector3 Vector3::kZero(0.0, 0.0, 0.0);
...
+                 const double m12, const double m20, const double m21, const double m22)
+    : Matrix3{{m00, m01, m02}, {m10, m11, m12}, {m20, m21, m22}} {}
+
+Matrix3::Matrix3() : Matrix3{{0.0, 0.0, 0.0}, {0.0, 0.0, 0.0}, {0.0, 0.0, 0.0}} {}
+
+Matrix3::Matrix3(const Matrix3& mat)
+    : Matrix3{mat[0][0], mat[0][1], mat[0][2], mat[1][0], mat[1][1], mat[1][2], mat[2][0], mat[2][1], mat[2][2]} {}
+
+Matrix3::Matrix3(Matrix3&& mat)
+    : Matrix3{mat[0][0], mat[0][1], mat[0][2], mat[1][0], mat[1][1], mat[1][2], mat[2][0], mat[2][1], mat[2][2]} {}
```
Comment (glpuga) [inline course/src/Isometry.cc]: You need to use move semantics, otherwise you're just creating a copy.

File: course/src/Isometry.cc [hunk 18]
```diff
@@ -96,21 +98,152 @@ const Vector3 Vector3::kZero(0.0, 0.0, 0.0);
...
 }
 
+bool Matrix3::operator==(const std::initializer_list<double>& rhs) const {
+    if (rhs.size() < 9) {
+        return false;
+    }
+    auto it = rhs.begin();
+
+    for (auto idx = 0; idx < 3; ++idx) {
+        if (mat_[idx] != std::initializer_list<double>({*it++, *it++, *it++})) {
```
Comment (glpuga) [inline course/src/Isometry.cc]: Look at this comment: https://github.com/LucasJSch/cppl1_q12020/pull/2#discussion_r381283578

**Review by glpuga — CHANGES_REQUESTED** [review]
Almost there. Review move semantics and look at the issue of order of evaluation of arguments.

**Review by glpuga — APPROVED** [review]
LGTM! Good work! Please squash, merge, and rebase your remaining PR on top of those changes.

**Pull request merged**